### ISSUE TITLE
CORS error after successful sign-in

### ISSUE THREAD

**Issue by Oliver15-dev** [MicrosoftDocs/msteams-docs#7641]
Hi everyone,
I am currently receiving a CORS error after a successful sign-in when I run an ajax call to my web app that’s installed on my server.
![image](https://user-images.githubusercontent.com/55416849/206301396-f838a58d-9752-44a9-921f-b48acee319d9.png)

This is preventing my configure page from rendering inside the Teams tab configuration section.

I have tried the following:

- Added app.UseCors(Microsoft.Owin.Cors.CorsOptions.AllowAll) to Startup.cs;
- Added the following headers to my ajax call as well as IIS:
'Access-Control-Allow-Origin': ‘login.microsoft.com,mydomain.com.au' (also tried ‘*’)

I have also tried the prerequisites on [this page](https://learn.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/tab-requirements) such as:
- Adding "frame-ancestors teams.microsoft.com *.teams.microsoft.com *.skype.com" to the Content-Security-Policy header in both IIS and my ASP.NET MVC app
- Adding X-Frame-Options: ALLOW-FROM https://teams.microsoft.com/ header in both IIS and my ASP.NET MVC app

I have spent some time researching the error but none of the fixes seems to work.
My web app is built using asp.net 4 MVC (version 4.7.2)

My app was working in October, so I am unsure if a Teams update caused this issue.

Has anyone seen this issue before, and if so, what was the fix?

Thank you,
Oliver

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: b159fe6e-33d0-6174-f3c3-68aaec79ee1b
* Version Independent ID: 7037e9c7-f501-5b70-8acf-1f004a06289b
* Content: [Prerequisites - Teams](https://learn.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/tab-requirements)
* Content Source: [msteams-platform/tabs/how-to/tab-requirements.md](https://github.com/MicrosoftDocs/msteams-docs/blob/main/msteams-platform/tabs/how-to/tab-requirements.md)
* Service: **msteams**
* GitHub Login: @surbhigupta
* Microsoft Alias: **lajanuar**

**Comment by Oliver15-dev** [comment]
@sayali-MSFT Thank you for the suggestions.

I have tried both enabling CORS via IIS 10 and in my ASP.NET Code via these two links:
https://enable-cors.org/server_aspnet.html
https://enable-cors.org/server_iis7.html

However I am still receiving the CORS errors above

I also have the IIS 10 CORS module installed
https://www.iis.net/downloads/microsoft/iis-cors-module

**Comment by sayali-MSFT** [comment]
@Oliver15-dev - The [user agent](https://developer.mozilla.org/en-US/docs/Glossary/User_agent) was unable to add the required [Origin](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin) header to the [HTTP](https://developer.mozilla.org/en-US/docs/Glossary/HTTP) request. All CORS requests must have an Origin header.

This can happen if the JavaScript code is running with enhanced privileges allowing it access to multiple domains' content, for example.

**Comment by Oliver15-dev** [comment]
@sayali-MSFT How can I tell if my javascript code is running with enhanced privileges? 

Also, in terms of the CORS headers, this is what I've done:

In the c#ASP.NET code I have added:
- To the **Startup.cs** file
```
app.UseCors(Microsoft.Owin.Cors.CorsOptions.AllowAll);
```

- To the **Global.asax** file:
```  var context = System.Web.HttpContext.Current;
     context.Response.AddHeader("Content-Security-Policy", "frame-ancestors teams.microsoft.com *.teams.microsoft.com *.skype.com");
     context.Response.AddHeader("Access-Control-Allow-Origin", "{mydomain}.com.au login.microsoftonline.com"); ```

- To my controllers
```
 [EnableCors(origins: "https://login.microsoftonline.com,{mydomain}.com.au", headers: "*", methods: "*")]
```

- To **WebApiConfig.cs** in the **Register** method
```
var corsAttr = new EnableCorsAttribute("https://login.microsoftonline.com,{mydomain}.com.au", "*", "*");
config.EnableCors(corsAttr);
```

and in IIS10:
![image](https://user-images.githubusercontent.com/55416849/206934030-9fdb7ea5-84a4-4ae3-8f1c-f70a722d34e2.png)


Am I missing anything/have I misconfigured anything?

**Comment by sayali-MSFT** [comment]
@Oliver15-dev -Follow the below 2 steps to enable CORS in your ASP.NET app:

1. Install the Microsoft.AspNetCore.Cors Nuget package.
2. Register CORS in the ConfigureService() method of Startup.cs. Next, enable CORS middleware in the Configure() method of Startup.cs.
Note: in .NET 6 or later versions, we need to perform 2nd step on Program.cs class.

We need to Register CORS as a service in the Startup.cs file’s ConfigureServices() method as shown below:
```
public void ConfigureServices(IServiceCollection services)
{
    services.AddCors();
    services.AddControllersWithViews();
}
```

For .NET 6 or later versions: The above code for .NET 6 or later version is added to Program class as shown below:

```
var builder = WebApplication.CreateBuilder(args); 
builder.Services.AddCors();

 // Shows UseCors with CorsPolicyBuilder.
app.UseCors(builder =>
{
    builder.AllowAnyOrigin()
           .AllowAnyMethod()
           .AllowAnyHeader();
});
```

**Comment by Oliver15-dev** [comment]
Hi @sayali-MSFT ,

Thank you for your suggestion, however, my App is built using ASP.NET 4.7, so it doesn't not have a ConfigureServices method

This is what I have in my Startup.cs file to enable CORS
![image](https://user-images.githubusercontent.com/55416849/207991508-dd163b73-26fb-4240-9fd7-0f32148b0054.png)

**Comment by sayali-MSFT** [comment]
@Oliver15-dev - Thanks u for the detailed information. We are checking it with internally and let you know the update.

**Comment by Oliver15-dev** [comment]
Thank you @sayali-MSFT! Did you manage to find anything?

**Comment by sayali-MSFT** [comment]
@Oliver15-dev - The error seems to imply that your app is making an AJAX request to some login.microsftonline.com endpoint from its own domain and AAD is not returning the right CORS header. You need to modify your app's registration in the AAD portal.

similar issue discussed here:
https://stackoverflow.com/questions/68585377/access-to-xmlhttprequest-at-https-login-microsoftonline-com-has-been-blocked-b

**Comment by Oliver15-dev** [comment]
@sayali-MSFT Thank you for the information, The tutorials seem to be for ASP.NET CORE and .NET 6, are there any tutorials for ASP.NET MVC (4.7.2)?

Also has there been a change to Azure that requires the changes listed here?
https://stackoverflow.com/questions/68585377/access-to-xmlhttprequest-at-https-login-microsoftonline-com-has-been-blocked-b

Only asking as my app has been working for sometime now until a couple months ago.

You are correct, my app is making an AJAX request.

The app is currently setup to use OIDC Authentication, this is the code I have for the Configuration function:
```
public void Configuration(IAppBuilder app)
{
    try
    {
        app.UseCors(Microsoft.Owin.Cors.CorsOptions.AllowAll);

        GlobalHost.Configuration.TransportConnectTimeout = TimeSpan.FromSeconds(10);
        app.MapSignalR();

        app.SetDefaultSignInAsAuthenticationType(CookieAuthenticationDefaults.AuthenticationType);

        app.UseCookieAuthentication(new CookieAuthenticationOptions()
        {
            CookieManager = new SystemWebCookieManager(),
        });

        // Configure Open ID Connect
        app.UseOpenIdConnectAuthentication(
        new OpenIdConnectAuthenticationOptions
        {
            ResponseType = OpenIdConnectResponseType.Code,
            ClientId = MsalAppBuilder.applicationID,
            Authority = MsalAppBuilder.authority,
            RedirectUri = MsalAppBuilder.redirectUri,
            PostLogoutRedirectUri = MsalAppBuilder.redirectUri,
            Scope = "openid email profile offline_access Files.ReadWrite.All User.Read",
            AuthenticationMode = AuthenticationMode.Active,
            TokenValidationParameters = new TokenValidationParameters()
            {
                ValidateIssuer = false
            },
            Notifications = new OpenIdConnectAuthenticationNotifications
            {
                SecurityTokenValidated = async n =>
                {
                    // Set persistent cookie, 
                    n.AuthenticationTicket.Properties.IsPersistent = true;
                    // and the expiration
                    n.AuthenticationTicket.Properties.ExpiresUtc = DateTime.Today.AddDays(2);
                },
                AuthenticationFailed = OnAuthenticationFailed,
                AuthorizationCodeReceived = OnAuthorizationCodeReceivedAsync,
            },
            UsePkce = false,
        });
        
        //IdentityModelEventSource.ShowPII = true;
    }
}

```

After the user successfully signs in, within the javascript file, there is a GET ajax call to retrieve the user's settings:

```
microsoftTeams.initialize();

microsoftTeams.getContext(async function (context) { 
$.ajax({
    url: rootPath + '/Admin/GetSettings',
    type: "GET",
    headers: {
        'Content-Security-Policy': 'frame-ancestors teams.microsoft.com *.teams.microsoft.com *.skype.com',
        'Access-Control-Allow-Origin': 'https://login.microsoftonline.com https://mydomain.com.au'
    },
    data: { channelId: context.channelId, teamsId: context.groupId, channelName: context.channelName, teamName: context.teamName },
    async: false,
    success: function (data) {
        // notify success

    },
    error: function (data) {
        console.log("error retrieving Teams settings");
        console.log(data);

    }
})

}
```

Previously, the above ajax call was successful and would return a JSON of the user's settings, however, now it just throws the CORS error instead

**Comment by Oliver15-dev** [comment]
Hi @sayali-MSFT , have you had a chance to review the above information? Let me know if you need me to provide any more information help diagnose the issue

**Comment by sayali-MSFT** [comment]
@Oliver15-dev -We are trying to repro the issue and also check with the internal team and let you know the results. Thanks!

**Comment by Oliver15-dev** [comment]
Hi @sayali-MSFT , unfortunately I won't be able to provide a Fiddler Trace at this stage

rootPath is just a variable containing the sitepath
e.g. rootPath = "https://example.com.au/Teams2CMCollaborate"

**Comment by sayali-MSFT** [comment]
@Oliver15-dev - We are not able to investigation further due to unavailability of network traces/logs. 
Could you please monitor this error and try to provide network trace when it reappears?

**Comment by sayali-MSFT** [comment]
@Oliver15-dev -Please send the details here- microsoftteamsdev@microsoft.com

**Comment by Oliver15-dev** [comment]
@sayali-MSFT Thank you, I have sent the .har network trace to [microsoftteamsdev@microsoft.com](mailto:microsoftteamsdev@microsoft.com)

**Comment by sayali-MSFT** [comment]
@Oliver15-dev -Have you sent mail with .har network trace?
because we did not got the mail.

**Comment by Oliver15-dev** [comment]
@sayali-MSFT Yes, I sent the email on the 19/01. I have just sent another one from the following email address: Oliver.Muthusami@fyb.com.au

Please let me know if you receive the email

**Comment by sayali-MSFT** [comment]
@Oliver15-dev -Thanks for sharing the details, we are checking it internally and let you know once we get any update.

**Comment by Oliver15-dev** [comment]
@sayali-MSFT no worries, thank you for letting me know

**Comment by sayali-MSFT** [comment]
@Oliver15-dev -Just want to confirm one thing, have you done below steps?
If you attempt to use the authorization code flow without setting up CORS for your redirect URI, you will see this error in the console:
`access to XMLHttpRequest at 'https://login.microsoftonline.com/common/v2.0/oauth2/token' from origin 'yourApp.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
`

To solve the issue, you need to do below steps: -

 go to their app registration in Azure Portal, switch to the Authentication tab, and then migrate their callback URL(s) to from the "Web" to "Single-page application" section:
 
![MicrosoftTeams-image (31)](https://user-images.githubusercontent.com/93527375/214492661-b454d666-8b9f-4010-ad55-a74f663ed3e0.png)

Reference document-https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow#redirect-uris-for-single-page-apps-spas

**Comment by Oliver15-dev** [comment]
@sayali-MSFT Thank you for the advice, when I migrate my callback URL from "Web" to "Single-page application" and sign in, I get this error:

![image](https://user-images.githubusercontent.com/55416849/214494793-70f848d4-03bf-4eb3-b80c-115aa1599120.png)

**Comment by Oliver15-dev** [comment]
hi @sayali-MSFT , have your team received my network trace file?

**Comment by sayali-MSFT** [comment]
@Oliver15-dev -Yes, we receive network trace file. We are checking this with engineering team.
Currently we don't have update on this. Once we get any update, we will inform you.

**Comment by Oliver15-dev** [comment]
@sayali-MSFT too easy, thank you for the update

**Comment by Oliver15-dev** [comment]
Hi @sayali-MSFT has there been any updates on the issue?

**Comment by sayali-MSFT** [comment]
@Oliver15-dev - Could you please post this query on below to the AAD support channels as well.
https://learn.microsoft.com/en-us/azure/active-directory/fundamentals/how-to-get-support

**Comment by Oliver15-dev** [comment]
@sayali-MSFT all done, I've just opened a support ticket

**Comment by Inventrax69** [comment]
@Oliver15-dev , Did you get a solution to this , we are also facing the same issues. Can you suggest a workaround if you pass thru this.

**Comment by Oliver15-dev** [comment]
@Inventrax69 nah, my issue got escalated to a Techincal Lead, however we closed the case as we no longer support the solution that had the CORS error
